sae/vm.go

```diff
@@ -1,28 +1,180 @@
+	}
+	return &VM{
+		config: c,
+		blocks: newSMap[common.Hash, *blocks.Block](),
+	}
+}
+
+// Init initializes the [VM], similarly to the `Initialize` method of
+// [snowcommon.VM].
+func (vm *VM) Init(
+	snowCtx *snow.Context,
+	hooks hook.Points,
+	chainConfig *params.ChainConfig,
+	db ethdb.Database,
+	triedbConfig *triedb.Config,
+	lastSynchronous *blocks.Block,
+) error {
+	vm.snowCtx = snowCtx
+	vm.hooks = hooks
+	vm.blocks.Store(lastSynchronous.Hash(), lastSynchronous)
+	vm.preference.Store(lastSynchronous)
+	vm.lastAccepted.Store(lastSynchronous)
+
+	vm.metrics = prometheus.NewRegistry()
+	if err := snowCtx.Metrics.Register("SAE", vm.metrics); err != nil {
+		return err
+	}
+
+	{ // ==========  Executor  ==========
+		exec, err := saexec.New(
+			lastSynchronous,
+			vm.blockSource,
+			chainConfig,
+			db,
+			triedbConfig,
+			hooks,
+			snowCtx.Log,
+		)
+		if err != nil {
+			return fmt.Errorf("saexec.New(...): %v", err)
+		}
+		vm.exec = exec
+		vm.toClose = append(vm.toClose, exec.Close)
+	}
+
+	{ // ==========  Mempool  ==========
+		bc := txgossip.NewBlockChain(vm.exec, vm.blockSource)
+		pools := []txpool.SubPool{
+			legacypool.New(vm.config.MempoolConfig, bc),
+		}
+		txPool, err := txpool.New(0, bc, pools)
+		if err != nil {
+			return fmt.Errorf("txpool.New(...): %v", err)
+		}
+		vm.toClose = append(vm.toClose, txPool.Close)
+
+		metrics, err := bloom.NewMetrics("mempool", vm.metrics)
+		if err != nil {
+			return err
+		}
+		conf := gossip.BloomSetConfig{Metrics: metrics}
+		pool, err := txgossip.NewSet(snowCtx.Log, txPool, conf)
+		if err != nil {
+			return err
+		}
+		vm.mempool = pool
+		vm.signalNewTxsToEngine()
+	}
+
+	return nil
+}
+
+// signalNewTxsToEngine subscribes to the [txpool.TxPool] to unblock a current
+// or future call to [VM.WaitForEvent]. At most one future call will be
+// unblocked through the use of a buffered channel. [VM.Shutdown] MUST be
+// called to release a goroutine started by this method.
+func (vm *VM) signalNewTxsToEngine() {
+	ch := make(chan core.NewTxsEvent)
+	sub := vm.mempool.Pool.SubscribeTransactions(ch, false /*reorgs but ignored by legacypool*/)
+	vm.toClose = append(vm.toClose, func() error {
+		defer close(ch)
+		sub.Unsubscribe()
+		return <-sub.Err() // guaranteed to be closed due to unsubscribing
+	})
+
+	vm.newTxs = make(chan struct{}, 1)
+	go func() {
+		defer close(vm.newTxs)
+		for range ch {
+			select {
+			case vm.newTxs <- struct{}{}:
+				_ = 0 // coverage visualisation
+			default:
+				_ = 0 // coverage visualization
+			}
+		}
+	}()
+}
 
 // WaitForEvent blocks until the VM needs to signal the consensus engine, and
 // then returns the message to send.
-func (vm *VM) WaitForEvent(ctx context.Context) (common.Message, error) {
-	return 0, errUnimplemented
+func (vm *VM) WaitForEvent(ctx context.Context) (snowcommon.Message, error) {
+	select {
+	case _, ok := <-vm.newTxs:
+		if !ok {
+			return 0, errors.New("VM closed")
+		}
+		return snowcommon.PendingTxs, nil
```

Comment:
This implementation is really not sufficient. Specifically, returning `snowcommon.PendingTxs` does not guarantee that `BuildBlock` will be called.

Additionally, it may have been possible that `BuildBlock` was called but didn't include all txs from the mempool (which would result in txs stuck in the mempool).

Should there be a TODO added here?

Other things to consider: The mempool may not have actually finished reorging when this function is called, so we should probably wait for any outstanding reorgs to finish.